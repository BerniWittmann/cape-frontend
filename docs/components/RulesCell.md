# RulesCell

A Cell which allows to edit the Attribute Cell of the Rules in the Context Factor. Allows to edit Strings, Booleans and Numbers.

## Props

<!-- @vuese:RulesCell:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|type|The type of the input field|`'Number'` / `'Boolean'` / `'String'`|`true`|-|
|value|The value of the input field|`String`|`false`|-|

<!-- @vuese:RulesCell:props:end -->


## Events

<!-- @vuese:RulesCell:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|change|Emitted on change of the value|The updated value|

<!-- @vuese:RulesCell:events:end -->


## Methods

<!-- @vuese:RulesCell:methods:start -->
|Method|Description|Parameters|
|---|---|---|
|switchNumberView|switches the view from a single number input to the double number input and back|-|
|updateValue|updates a value change, checks string for forbidden characters (=, &, |) and allows to empty a boolean choice|the input/selection|

<!-- @vuese:RulesCell:methods:end -->


