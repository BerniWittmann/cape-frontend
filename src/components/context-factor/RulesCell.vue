<template>
  <div id="attribute_cell">
    <div v-if="type === 'Boolean'">
      <el-select v-model="data.text" class="fill" @change="updateValue" clearable ref="select">
        <el-option
                v-for="item in booleanOptions"
                :key="item.value"
                :label="item.value"
                :value="item.value">
        </el-option>
        <el-option
                :value="$t('context_factor.none')"
                :label="$t('context_factor.none')"></el-option>
      </el-select>
    </div>
    <div v-if="type === 'String'">
      <el-input v-model="data.text" clearable class="fill" @change="updateValue"></el-input>
    </div>
    <div v-if="type === 'Number' && singleNumberMode" class="n-in">
      <el-button icon="el-icon-d-caret" plain @click="switchNumberView" size="mini" class="switchButton"></el-button>
      <el-input-number v-model="integerValue1" clearable class="n_input fill" @change="updateValue"
                       :controls="false"></el-input-number>
    </div>
    <div v-if="type === 'Number' && !singleNumberMode" class="n-in">
      <el-button icon="el-icon-d-caret" plain @click="switchNumberView" size="mini" class="switchButton"></el-button>
      <el-input-number v-model="integerValue1" clearable class="n-input" @change="updateValue"
                       :controls="false"></el-input-number>
      <span class="separator"> - </span>
      <el-input-number v-model="integerValue2" clearable class="n-input" @change="updateValue"
                       :controls="false"></el-input-number>
    </div>
  </div>
</template>

<script>
/* ============
 * RulesCell
 * ============
 *
 * A Cell which allows to edit the Attribute Cell of the Rules in the Context Factor
 *
 */
export default {
  name: 'RulesCell',
  model: {
    prop: 'value',
    event: 'change'
  },

  props: {
    // The Value of the input field
    type: {
      type: String,
      required: true
    },
    value: {
      type: String
    }
  },

  data() {
    return {
      booleanOptions: [{
        value: 'TRUE'
      }, {
        value: 'FALSE'
      }],
      data: {
        text: ''
      }
    }
  },

  computed: {
    singleNumberMode() {
      return this.data.text === undefined || this.data.text.indexOf('-') === -1
    },

    integerValue1: {
      get: function () {
        if (this.type === 'Number' && !this.singleNumberMode) {
          return this.data.text.split('-')[0].length > 0 ? this.data.text.split('-')[0] : undefined
        }
        if (this.type === 'Number' && this.singleNumberMode) {
          return this.data.text
        }
        return undefined
      },
      set: function (newInteger1Value) {
        if (this.singleNumberMode) {
          this.data.text = newInteger1Value + ''
        } else {
          if (newInteger1Value !== undefined) {
            if (newInteger1Value > this.integerValue2 && this.integerValue2 !== undefined) {
              let temp = this.integerValue2
              this.integerValue2 = newInteger1Value
              newInteger1Value = temp
              this.data.text = this.data.text.split('-')[0] + '-' + this.integerValue2
            }
            if (this.data.text.indexOf('-') === 0) {
              this.data.text = newInteger1Value + this.data.text
            } else {
              this.data.text = newInteger1Value + '-' + this.data.text.split('-')[1]
            }
          }
        }
      }
    },

    integerValue2: {
      get: function () {
        if (this.type === 'Number' && !this.singleNumberMode) {
          return this.data.text.split('-')[1].length > 0 ? this.data.text.split('-')[1] : undefined
        }
        return undefined
      },
      set: function (newInteger2Value) {
        if (newInteger2Value !== undefined) {
          if (newInteger2Value < this.integerValue1) {
            let temp = this.integerValue1
            this.integerValue1 = newInteger2Value
            newInteger2Value = temp
          }
          if (this.data.text.indexOf('-') === this.data.text.length - 1) {
            this.data.text += newInteger2Value
          } else {
            this.data.text = this.data.text.split('-')[0] + '-' + newInteger2Value
          }
        }
      }
    }
  },

  methods: {
    switchNumberView() {
      // the separator for the numbers must be added or removed
      if (this.singleNumberMode) {
        this.data.text += '-'
      } else {
        this.data.text = this.data.text.split('-')[0]
      }
    },

    updateValue(selected) {
      // check if boolean input value is none to remove the text and attribute
      if (selected === this.$t('context_factor.none')) this.data.text = ''
      this.$emit('change', this.data.text)
    }
  },

  watch: {
    value: {
      deep: true,
      immediate: true,
      handler: function (newValue) {
        if (this.type === 'Boolean' && newValue !== 'TRUE' && newValue !== 'FALSE') newValue = undefined
        this.data.text = newValue
        this.$emit('change', this.data.text)
      }
    }
  }
}
</script>

<style scoped lang="scss">

.switchButton {
  padding: 2px;
  margin-right: 5px;
}

.separator {
  margin: 5px;
}

.n-input {
  flex: 1
}

.n-in {
  display: flex;
}

.fill {
  width: 100%;
}

</style>
