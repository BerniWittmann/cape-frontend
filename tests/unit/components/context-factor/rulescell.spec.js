import { shallowMount } from '@vue/test-utils'
import { i18n } from '../../setupPlugins'

import RulesCell from '@/components/context-factor/RulesCell.vue'

describe('Components', () => {
  describe('RulesCell', () => {
    let cmp
    let props
    beforeEach(() => {
      jest.clearAllMocks()
      props = {
        type: 'String'
      }
      render()
    })

    function render() {
      cmp = shallowMount(RulesCell, {
        i18n,
        propsData: props
      })
    }

    describe('its of type String', () => {
      beforeEach(() => {
        props.type = 'String'
        cmp.setProps(props)
      })
      it('renders', () => {
        expect(cmp.html()).toMatchSnapshot()
      })
      it('renders an input field', () => {
        const input = cmp.find('elinput-stub')
        expect(input.exists()).toBeTruthy()
      })
      it('updates the value on change', () => {
        cmp.vm.data.text = 'My Value'
        cmp.find('elinput-stub').vm.$emit('change')
        expect(cmp.emitted().change).toBeTruthy()
        expect(cmp.emitted().change[1][0]).toEqual('My Value')
      })
      it('integerValues have no value', () => {
        expect(cmp.vm.integerValue1).toBeUndefined()
        expect(cmp.vm.integerValue2).toBeUndefined()
      })
    })

    describe('its of type Number', () => {
      beforeEach(() => {
        props.type = 'Number'
        cmp.setProps(props)
      })
      it('renders', () => {
        expect(cmp.html()).toMatchSnapshot()
      })
      it('renders an input field', () => {
        const input = cmp.find('elinputnumber-stub')
        expect(input.exists()).toBeTruthy()
      })
      it('updates the value on change', () => {
        cmp.vm.data.text = '12'
        cmp.find('elinputnumber-stub').vm.$emit('change')
        expect(cmp.emitted().change).toBeTruthy()
        expect(cmp.emitted().change[1][0]).toEqual('12')
      })
      it('can switch number mode', () => {
        expect(cmp.findAll('elinputnumber-stub').length).toEqual(1)
        const btn = cmp.find('.switchButton')
        expect(btn.exists()).toBeTruthy()
        btn.vm.$emit('click')
        expect(cmp.html()).toMatchSnapshot()
        expect(cmp.findAll('elinputnumber-stub').length).toEqual(2)
      })
      describe('is in interval mode', () => {
        beforeEach(() => {
          const btn = cmp.find('.switchButton')
          expect(btn.exists()).toBeTruthy()
          btn.vm.$emit('click')
          expect(cmp.findAll('elinputnumber-stub').length).toEqual(2)
        })
        it('correctly displays the interval', () => {
          cmp.vm.data.text = '14_15'
          expect(cmp.vm.integerValue1).toEqual('14')
          expect(cmp.vm.integerValue2).toEqual('15')
        })
        it('handles invalid text', () => {
          cmp.vm.data.text = '14'
          expect(cmp.vm.integerValue1).toEqual('14')
          expect(cmp.vm.integerValue2).toBeUndefined()

          cmp.vm.integerValue1 = '12'
          cmp.vm.integerValue2 = '99'
          expect(cmp.vm.data.text).toEqual('12_99')

          cmp.vm.data.text = '_'
          expect(cmp.vm.integerValue1).toBeUndefined()
          expect(cmp.vm.integerValue2).toBeUndefined()
        })
        it('handles undefined values', () => {
          cmp.vm.integerValue1 = undefined
          cmp.vm.integerValue2 = undefined
          expect(cmp.vm.data.text).toEqual('_')

          cmp.vm.data.text = '_'
          cmp.vm.integerValue1 = '99'
          expect(cmp.vm.data.text).toEqual('99_')
        })
        it('can switch number mode', () => {
          const btn = cmp.find('.switchButton')
          expect(btn.exists()).toBeTruthy()
          btn.vm.$emit('click')
          expect(cmp.html()).toMatchSnapshot()
          expect(cmp.findAll('elinputnumber-stub').length).toEqual(1)
        })
        it('updates the value on change', () => {
          cmp.vm.integerValue1 = 12
          cmp.vm.integerValue2 = 23
          cmp.find('elinputnumber-stub').vm.$emit('change')
          expect(cmp.emitted().change).toBeTruthy()
          expect(cmp.emitted().change[1][0]).toEqual('12_23')
        })
        it('corrects the interval border ordering', () => {
          cmp.vm.integerValue2 = 30
          cmp.vm.integerValue1 = 90
          expect(cmp.vm.data.text).toEqual('30_90')
        })
        it('corrects the interval border ordering', () => {
          cmp.vm.integerValue1 = 99
          cmp.vm.integerValue2 = 30
          cmp.vm.integerValue1 = 1
          cmp.find('elinputnumber-stub').vm.$emit('change')
          expect(cmp.emitted().change).toBeTruthy()
          expect(cmp.emitted().change[1][0]).toEqual('1_99')
        })
      })
    })

    describe('its of type Boolean', () => {
      beforeEach(() => {
        props.type = 'Boolean'
        cmp.setProps(props)
      })
      it('renders', () => {
        expect(cmp.html()).toMatchSnapshot()
      })
      it('renders a select', () => {
        const select = cmp.find('elselect-stub')
        expect(select.exists()).toBeTruthy()
        const options = select.findAll('eloption-stub')
        expect(options.length).toEqual(3)
        expect(options.at(0).attributes('value')).toEqual('TRUE')
        expect(options.at(1).attributes('value')).toEqual('FALSE')
        expect(options.at(2).attributes('value')).toEqual('context_factor.none')
      })
      it('updates the value on change', () => {
        cmp.vm.data.text = 'TRUE'
        cmp.find('elselect-stub').vm.$emit('change')
        expect(cmp.emitted().change).toBeTruthy()
        expect(cmp.emitted().change[1][0]).toEqual('TRUE')
      })
      it('updates the empty value on change', () => {
        cmp.vm.data.text = 'TRUE'
        cmp.find('elselect-stub').vm.$emit('change', 'context_factor.none')
        expect(cmp.emitted().change).toBeTruthy()
        expect(cmp.emitted().change[1][0]).toEqual('')
      })
      it('handles an invalid value update', () => {
        cmp.setProps({ value: 'invalid' })
        expect(cmp.vm.data.text).toBeUndefined()
        expect(cmp.emitted().change).toBeTruthy()
        expect(cmp.emitted().change[1][0]).toEqual(undefined)
      })
    })
  })
})
