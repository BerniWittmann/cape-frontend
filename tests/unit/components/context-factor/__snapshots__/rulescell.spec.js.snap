// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components RulesCell its of type Boolean renders 1`] = `
<div id="attribute_cell">
  <div>
    <elselect-stub clearable="true" multiplelimit="0" placeholder="请选择" valuekey="value" popperappendtobody="true" class="fill">
      <eloption-stub value="TRUE" label="TRUE"></eloption-stub>
      <eloption-stub value="FALSE" label="FALSE"></eloption-stub>
      <eloption-stub value="context_factor.none" label="context_factor.none"></eloption-stub>
    </elselect-stub>
  </div>
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`Components RulesCell its of type Number can switch number mode 1`] = `
<div id="attribute_cell">
  <!---->
  <!---->
  <!---->
  <div class="n-in">
    <elbutton-stub type="default" size="mini" icon="el-icon-d-caret" nativetype="button" plain="true" class="switchButton"></elbutton-stub>
    <elinputnumber-stub step="1" max="Infinity" min="-Infinity" controlsposition="" clearable="" class="n-input" value="undefined"></elinputnumber-stub> <span class="separator"> - </span>
    <elinputnumber-stub step="1" max="Infinity" min="-Infinity" controlsposition="" clearable="" class="n-input"></elinputnumber-stub>
  </div>
</div>
`;

exports[`Components RulesCell its of type Number is in interval mode can switch number mode 1`] = `
<div id="attribute_cell">
  <!---->
  <!---->
  <div class="n-in">
    <elbutton-stub type="default" size="mini" icon="el-icon-d-caret" nativetype="button" plain="true" class="switchButton"></elbutton-stub>
    <elinputnumber-stub step="1" max="Infinity" min="-Infinity" controlsposition="" clearable="" class="n_input fill" value="undefined"></elinputnumber-stub>
  </div>
  <!---->
</div>
`;

exports[`Components RulesCell its of type Number renders 1`] = `
<div id="attribute_cell">
  <!---->
  <!---->
  <div class="n-in">
    <elbutton-stub type="default" size="mini" icon="el-icon-d-caret" nativetype="button" plain="true" class="switchButton"></elbutton-stub>
    <elinputnumber-stub step="1" max="Infinity" min="-Infinity" controlsposition="" clearable="" class="n_input fill"></elinputnumber-stub>
  </div>
  <!---->
</div>
`;

exports[`Components RulesCell its of type String renders 1`] = `
<div id="attribute_cell">
  <!---->
  <div>
    <elinput-stub type="text" autocomplete="off" validateevent="true" clearable="true" class="fill"></elinput-stub>
  </div>
  <!---->
  <!---->
</div>
`;
